models/HubSpotHttpError.ts

```diff
@@ -1,41 +1,239 @@
   public statusText?: string;
   public data?: T;
-  public headers?: RawAxiosResponseHeaders | AxiosResponseHeaders;
+  public headers?: { [key: string]: unknown };
+  public method: string | undefined;
+  public context: HubSpotHttpErrorContext | undefined;
+  public validationErrors: string[] | undefined;
 
-  constructor(message?: string, options?: ErrorOptions) {
+  constructor(
+    message?: string,
+    options?: ErrorOptions,
+    context?: HubSpotHttpErrorContext
+  ) {
     super(message, options);
     this.name = 'HubSpotHttpError';
+    this.context = context;
 
     if (options && isAxiosError(options.cause)) {
-      const { response, request, config, code } = options.cause;
-      this.response = response;
+      this.updateContextFromCause(options.cause, context);
+      const { response, config, code } = options.cause;
+      this.joinErrorMessages(options.cause, context);
 
-      this.config = config;
       this.code = code;
-      this.request = request;
+      this.method = config?.method;
 
       // Pull the request fields to the top level
       if (response) {
         this.status = response.status;
         this.statusText = response.statusText;
         this.data = response.data;
         this.headers = response.headers;
+        this.parseValidationErrors(response.data);
       }
+    } else if (options && options.cause instanceof Error) {
+      const messages =
+        options.cause.name !== 'Error' ? [`${options.cause.name}:`] : [];
+
+      if (options.cause.message) {
+        messages.push(options.cause.message);
+      }
+
+      if ('reason' in options.cause && options.cause.reason) {
+        messages.push(`${options.cause.reason}`);
+      }
+
+      this.message = messages.join(' ');
+    }
+  }
+
+  public updateContext(context: Partial<HubSpotHttpErrorContext>) {
+    this.context = { ...this.context, ...context };
+  }
+
+  public toString() {
+    let baseString = `${this.name}: \n- message: ${this.message}`;
+    if (this.validationErrors && this.validationErrors.length > 0) {
+      baseString = `${baseString} \n- errors: ${this.validationErrors.join('\n- ')}`;
+    }
+    if (this.context) {
+      baseString = `${baseString} \n- context: ${JSON.stringify(this.context, undefined, 2)}`;
+    }
+    return baseString;
+  }
+
+  private updateContextFromCause(
+    cause: AxiosError,
+    context?: HubSpotHttpErrorContext
+  ) {
+    const generatedContext: HubSpotHttpErrorContext = {};
+    if (!cause) {
+      return;
+    }
+
+    try {
+      generatedContext.accountId = cause.config?.params?.portalId;
+      generatedContext.payload = JSON.stringify(cause.config?.data);
+      // This will just be the url path
+      generatedContext.request = cause.config?.url;
+    } catch (e) {
+      logger.debug(e);
+    }
+
+    // Allow the provided context to override the generated context
+    this.context = { ...generatedContext, ...context };
+  }
+
+  private parseValidationErrors(
```

Comment:
I think we might need to dig into this parse validation util. I noticed that we're logging out some of the important context, but not all of it. For example:

Here's what logs out when I get a "missing scopes" error response. I can get this by trying to run this command against my developer portal, which doesn't have the right content scopes to access CMS data.
<img width="853" alt="image" src="https://github.com/user-attachments/assets/7b62c13a-934d-4a81-8c64-a8880e0bd588">

It's trying to tell me the scopes I'm missing, but it's not including all of the information that I need. This is the shape of `responseData` when that happens:
<img width="844" alt="image" src="https://github.com/user-attachments/assets/ef74be07-7fc8-441c-809f-b8e51206add6">

`context`  here contains the scopes, like this: `["content", "content"]` (not sure why it's in there twice)

---

Also, from what I can tell, in the previous version of the error handling [we only called it in a few places ](https://github.com/search?q=repo%3AHubSpot%2Fhubspot-local-dev-lib%20throwApiUploadError&type=code) (throwApiUploadError is the only util that called isApiUploadValidationError,  which ran this validation parsing logic). I wonder if it will have an impact on some of the existing error messages?

Reply:
I added a method to log the [formattedValidationErrors](https://github.com/HubSpot/hubspot-local-dev-lib/pull/159/files#diff-ae25cee36b980aa393de912585da6d585d226617b9595f4a65e6e3215c2a1925R93-R98) that the [CLI now calls if the error is a validation error.](https://github.com/HubSpot/hubspot-cli/pull/1109/files#diff-7fef93fbb06f6ac57ebe30f3ac8e8ab4fd9ececfbf84c740d5a255d1ff1c1cb2R25-R27)

I'll look into using the context object.  How did you trigger missing scopes error?  Just remove those scopes from the app?

Reply:
The reference to "app" here is misleading. In reality the error is referencing the personal access key. So you can generate a personal access key that doesn't have any scopes, and then try to run various CLI commands.